src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -373,6 +374,161 @@ void optimizeOrderBy(const ASTSelectQuery * select_query)
         elems = std::move(unique_elems);
 }
 
+/// Checks if ASTFunction or its arguments are stateful.
+bool isASTFunctionStateful(const ASTPtr & current_ast, const Context & context)
+{
+    if (!current_ast)
+        return false;
+
+    if (auto ast_function = current_ast->as<ASTFunction>())
+    {
+        if (ast_function->name == "any" || ast_function->name == "groupArray")
+            return true;
+
+        const auto & function = FunctionFactory::instance().tryGet(ast_function->name, context);
+
+        if (function && function->isStateful())
+            return true;
+
+        for (const auto & elem : ast_function->children)
+        {
+            if (isASTFunctionStateful(elem, context))
+                return true;
+        }
+    }
+    else if (auto ast_expression_list = current_ast->as<ASTExpressionList>())
+    {
+        for (const auto & elem : ast_expression_list->children)
+        {
+            if (isASTFunctionStateful(elem, context))
+                return true;
+        }
+    }
+
+    return false;
+}
+
+/// Removes duplicate ORDER BY from subqueries.
+void optimizeDuplicateOrderByFromSubqueries(const ASTPtr & current_ast, const Context & context)
```

Comment:
We have generalized AST traversal described in InDephNodeVisitor. The primitives go through childrena and if can mach type they call matcher function for the type. You chould not recurse to children yourself. Use OneTypeMatcher with ASTSelectQuery as type and make Visitor based on it.